py_trees/decorators.py

```diff
@@ -0,0 +1,457 @@
+            new_status = common.Status.INVALID
+        if new_status != common.Status.RUNNING:
+            self.stop(new_status)
+        self.status = new_status
+        yield self
+
+    def stop(self, new_status):
+        """
+        As with other composites, it checks if the child is running
+        and stops it if that is the case. 
+
+        Args:
+            new_status (:class:`~py_trees.common.Status`): the behaviour is transitioning to this new status
+        """
+        self.logger.debug("%s.stop(%s)" % (self.__class__.__name__, new_status))
+        self.terminate(new_status)
+        # priority interrupt handling
+        if new_status == common.Status.INVALID:
+            self.decorated.stop(new_status)
+        # if the decorator returns SUCCESS/FAILURE and should stop the child
+        if self.decorated.status == common.Status.RUNNING:
+            self.decorated.stop(new_status)
+        self.status = new_status
+ 
+##############################################################################
+# Decorators
+##############################################################################
+
+class Timeout(Decorator):
+    """
+    A decorator that applies a timeout pattern to an existing behaviour.
+    If the timeout is reached, the encapsulated behaviour's
+    :meth:`~py_trees.behaviour.Behaviour.stop` method is called with
+    status :data:`~py_trees.common.Status.FAILURE` otherwise it will
+    simply directly tick and return with the same status
+    as that of it's encapsulated behaviour.
+    """
+    def __init__(self,
+                 child,
+                 name=common.Name.AUTO_GENERATED,
+                 duration=5.0):
+        """
+        Init with the decorated child and a timeout duration.
+                
+        Args:
+            child (:class:`~py_trees.behaviour.Behaviour`): behaviour to time
+            name (:obj:`str`): the decorator name
+            duration (:obj:`float`): timeout length in seconds
+        """
+        super(Timeout, self).__init__(name=name, child=child)
+        self.duration = duration
+        self.finish_time = None
+
+    def initialise(self):
+        """
+        Reset the feedback message and finish time on behaviour entry.
+        """
+        self.finish_time = time.time() + self.duration
+        self.feedback_message = ""
+ 
+    def update(self):
+        """
+        Terminate the child and return :data:`~py_trees.common.Status.FAILURE`
+        if the timeout is exceeded.
+        """
+        current_time = time.time()
+        if current_time > self.finish_time:
+            self.feedback_message = "timed out"
+            self.logger.debug("{}.update() {}".format(self.__class__.__name__, self.feedback_message))
+            # invalidate the decorated (i.e. cancel it), could also put this logic in a terminate() method
+            self.decorated.stop(common.Status.INVALID)
+            return common.Status.FAILURE
```

Comment:
this means that `stop` on decorated will be called twice. Once with this invoke here (with INVALID) and other in `tick` after we return FAILURE (with FAILURE).

Reply:
Aye.